text/0000-env-sandboxing.md

```diff
@@ -0,0 +1,289 @@
+to control which environment variables are accessible, or what their apparent
+contents are.
+
+This poses a few problems:
+
+- The build system has no idea what variables the code is using for the purposes
+  of tracking its inputs
+- Code can read arbitrary contents from arbitrary variables, which may pose
+  unwanted information leakage (from build environment to final deployment
+  environment, for example).
+- It combines the environment `rustc` needs for its own operation (PATH and
+  LD_LIBRARY_PATH, for example) with environment the compiled code might want,
+  and doesn't allow them to be set independently.
+
+## Include Files
+Rust allows any file to be directly included to be logically part of the
+including source file, either as more Rust source code (include!()), a text
+string (include_str!()) or arbitrary binary data (include_bytes!()). These
+macros take a raw string which is used as an arbitrary path which may be
+absolute - they therefore allow any file that `rustc` has permission to access
+to be used in the compiled output.
+
+(This differs from separating a crate into multiple source files via modules, as
+those files are always relative to the top-level lib.rs/main.rs source.)
+
+This causes a couple of problems:
+
+- The build system can't know or constrain what files are actually inputs to the
+  compiler.
+- The code can't unintentionally leak state from the build environment to the
+  deployment environment.
+
+## Note
+This RFC specifically does not intend to address any actual security concerns.
+There are many other avenues that it does not attempt to control, such as
+compiler plugins/proc macros. However, it does help with unintentional problems
+which could result in later security problems if unaddressed. Blocking access to
+files or environment variables also makes it easier to audit what environment
+accesses a codebase is attempting to make.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Environment Variables
+
+Rust supports the `env!()` and `option_env!()` macros to allow the value of a
+process environment variable to be logically pasted into the source as a string
+(or `Option` string).
+
+`Rustc` has the following command-line options to allow the access to
+environment variables to be controlled at a fine-grain level:
+
+- `--env-clear` - completely empty the logical environment visible to
+   `env!()`/`option_env!()`, causing them to all fail/return `None`. Without any
+   other options, this will completely disable environment variable access.
+- `--env-allow NAME` - allow a specific environment variable to be read from the
+  process environment.
+- `--env-define NAME=VAL` - define a logical environment variable. This does not
+  need to be present in the actual process environment, or if it is, its value
+  is overridden.
+
+These are processed in the order:
+1. `--env-clear`
+2. `--env-allow NAME`
+3. `--env-set NAME=VAL`
```

Comment:
what's the benefit of this (`rustc --env-set NAME=value`) over `NAME=value rustcy --env-allow NAME` other than brevity? is it specifically to stop propagation into sub-processes?

Reply:
Several reasons:
- You can use it to override the value of things that rustc (or the linker) itself uses, like `PATH` or `LD_LIBRARY_PATH` (among others)
- You can just change `rustc`'s command-line args, whereas manipulating the environment might need deeper changes to whatever's invoking `rustc` (possibly system-dependent changes)
- Non-propagation isn't something that had occurred to me, but it is related to my first point

Reply:
my question was more that my understanding is that running `> NAME=value rustc --env-allow NAME` *is* only overriding the value for that instantiation of `rustc`, so I don't understand what your first point is specifying? are you meaning that some point before/around your 'sandboxed' area is able to access the environment but then the majority is going to use the `--env-allow` value?

your option may also be more sensible for any shells/clients which don't support 'instantiation-local environment variables', but I think the RFC could do with more discussion of these points.

thanks!

Reply:
I mean that when `rustc` itself is - say - invoking the linker, it may use `$PATH` to find the executable. If we want `env!("PATH")` to work at all, but not necessarily return the `PATH` that's been set up for `rustc` to be able to find its linker, then we need some way of allowing `env!("PATH")` and `rustc`'s own runtime call to `std::env::var("PATH")` to be able to return distinct values.